139550-list-total-processed-payment-requests

simonnineteen90 opened the pull request:
https://dev.azure.com/defragovuk/DEFRA-FFC/_workitems/edit/139550

simonnineteen90 commented on test/unit/data/events/scheme/get-submitted-events.test.js:
```diff
@@ -0,0 +1,97 @@
+    mockListEntities.mockReturnValue([])
+    const result = await getSubmittedEvents(PARTITION_KEY, CATEGORY)
+    expect(result.length).toBe(0)
+  })
```
Is there a way to test the mock table storage as explain in comments here?
Can you explain what exactly the imported `0data` /`mock0Data` is.

johnwatson484 commented on test/unit/data/events/scheme/get-submitted-events.test.js:
```diff
@@ -0,0 +1,97 @@
+  })
+
+  // can this be tested because we are mocking the response from table storage?
+  test('should only return events that are of type submitted', async () => {
+    const result = await getSubmittedEvents(PARTITION_KEY, CATEGORY)
+    result.forEach(event => {
+      expect(event.type).toBe(PAYMENT_SUBMITTED)
+    })
+  })
```
Not with mocks in unit tests.  These tests are more testing the interface for table storage is being used as expected with the correct query parameters and the responses are handled appropriately. 

The type of test you're describing would be a local integration test in the `integration/local` folder.   You could load some events to storage, then use the SDK without mocks to see if only the submitted events are returned.

simonnineteen90 replied:
Integration test added.

johnwatson484 commented on test/unit/data/events/scheme/get-submitted-events.test.js:
```diff
@@ -0,0 +1,97 @@
+  // test that it is called with the correct type
+  test('should call mockListEntities with type of submitted ', async () => {
+    await getSubmittedEvents(PARTITION_KEY, CATEGORY)
+    // what is the mock0data doing and where does it come from
+    expect(mockTableClient.listEntities).toHaveBeenCalledWith({ queryOptions: { filter: mockOdata`THIS CAN BE ANYTHING AND IT WONT FAIL${PAYMENT_SUBMITTED}` } })
```
For more detail on odata you can read through this -> https://www.odata.org/ but as a summary it's a protocol for how to query data.

Table storage uses it and they SDK contains a handy helper function to parse js into it.

Hadn't appreciated the awkwardness of testing with it for this scenario.  I'd say make this test just confirm that the query is called correctly and create a local integration test to cover the actual usage.

I'm not sure why there aren't any local integration tests already.  Suspect an oversight.

simonnineteen90 commented on test/unit/data/events/scheme/get-events-by-scheme-id.test.js:
```diff
@@ -0,0 +1,57 @@
+
+describe('get events by frn', () => {
+  beforeEach(() => {
+    jest.clearAllMocks()
```
removing this `jest.clearAllMocks()` allows all tests to pass. Currently they all fail. Where is the correct place to rest the mocks or do i not need to in this scenario?

simonnineteen90 replied:
needed an `await` on each test

Kirsten1822 replied:
So with the clearAllMocks the tests fail?

simonnineteen90 replied:
no, they pass now i had missed off `await` when calling the function `getEventsBySchemeId` in the individual tests. I make this mistake roughly once a month 😅

Kirsten1822 replied:
You're fine!

Kirsten1822 commented on app/data/events/scheme-id/get-submitted-events.js:
```diff
@@ -0,0 +1,18 @@
+
+const getSubmittedEvents = async (category) => {
+  const client = getClient(PAYMENT_EVENT)
+  const eventResults = client.listEntities({ queryOptions: { filter: odata`category eq '${category}' and type eq '${PAYMENT_SUBMITTED}'` } })
```
Could you explain what this line of code does?

simonnineteen90 replied:
`client` is defined on line 6 and is getting the payment instance of the Azure Table Storage, this `client` has different functions we can call to interact with the data that is stored in the table.

line 7 is using the `listEntities` function to get all the 'entities' (which in our case are the events). We are then passing in an object which contains query data to filter the events. Specifically we are telling it to filter and only return the events that have a category of category (`schemeId`) and type of payment_submitted (`uk.gov.defra.ffc.pay.payment.submitted`)

So basically we use this line to get the events from the table storage based on the filters we pass in.

Kirsten1822 replied:
Thanks!

Kirsten1822 commented on test/integration/local/data/events/scheme/get-events-by-scheme-id.test.js:
```diff
@@ -0,0 +1,135 @@
+const { FRN } = require('../../../../../mocks/values/frn')
+const { INVOICE_NUMBER } = require('../../../../../mocks/values/invoice-number')
+const { BPS, CS, SFI } = require('../../../../../../app/constants/schemes')
```
I'd put a space here to separate the mocks from the constants

Kirsten1822 commented on test/integration/local/data/events/scheme/get-events-by-scheme-id.test.js:
```diff
@@ -0,0 +1,135 @@
+const { FRN } = require('../../../../../mocks/values/frn')
+const { INVOICE_NUMBER } = require('../../../../../mocks/values/invoice-number')
+const { BPS, CS, SFI } = require('../../../../../../app/constants/schemes')
+const schemeNames = require('../../../../../../app/constants/scheme-names')
+
+const { PAYMENT_EVENT, HOLD_EVENT, BATCH_EVENT, WARNING_EVENT } = require('../../../../../../app/constants/event-types')
```
These should probably be grouped with the other constants

Kirsten1822 commented on test/mocks/events/total-scheme-values.js:
```diff
@@ -0,0 +1,7 @@
+const { SCHEME_ID } = require('../values/scheme-id')
+const { VALUE } = require('../values/value')
+module.exports = {
```
Space needed here

simonnineteen90 replied:
also moved this out of the `events` folder as its not an event, doesnt really fit the other folders so put it in the root `mocks` folder

Kirsten1822 commented on test/unit/data/events/scheme/get-events-by-scheme-id.test.js:
```diff
@@ -0,0 +1,57 @@
+const { sanitiseSchemeData: mockSanitiseSchemeData } = require('../../../../../app/data/events/scheme-id/sanitise-scheme-data')
+
+const { SCHEME_ID: SCHEME_ID_CATEGORY } = require('../../../../../app/constants/categories')
+const submitted = require('../../../../mocks/events/submitted')
+const groupedEvent = require('../../../../mocks/events/grouped-event')
+const totalSchemeValues = require('../../../../mocks/events/total-scheme-values')
```
I'd put these mocks above the constant, with a space to separate them

Kirsten1822 commented on test/unit/data/events/scheme/get-submitted-events.test.js:
```diff
@@ -0,0 +1,86 @@
+const { getSubmittedEvents } = require('../../../../../app/data/events/scheme-id/get-submitted-events')
+
+const category = 'schemeId'
+let extractedEvent
```
Space needed here

Kirsten1822 commented on test/unit/data/events/scheme/get-total-scheme-values.test.js:
```diff
@@ -0,0 +1,76 @@
+const { getTotalSchemeValues } = require('../../../../../app/data/events/scheme-id/get-total-scheme-values')
+const { BPS, CS } = require('../../../../../app/constants/schemes')
```
Constant should be above getTotalSchemeValue - with a space to separate them

Kirsten1822 commented on test/unit/data/events/scheme/get-total-scheme-values.test.js:
```diff
@@ -0,0 +1,76 @@
+    expect(result[0].schemeId).toBe(groupedEvents[0].schemeId)
+  })
+
+  test('value should be equal to sum of events values when there are 3 events', () => {
```
Why only 3 events?

Kirsten1822 commented on test/unit/data/events/scheme/group-events-by-scheme.test.js:
```diff
@@ -0,0 +1,34 @@
+const { BPS, CS, SFI } = require('../../../../../app/constants/schemes')
+const submitted = require('../../../../mocks/events/submitted')
```
I'd put the mock at the top of the file with a space between to separate it from the constant

Kirsten1822 commented on test/unit/data/events/scheme/group-events-by-scheme.test.js:
```diff
@@ -0,0 +1,34 @@
+    mixedSchemeEvents = [...bpsEvents, ...csEvents, ...sfiEvents]
+  })
+
+  test('returned schemeId should be equal to partitionKey of all events in eventGroup', () => {
```
Would 'should return schemeId equal to partitionKey of all events in eventGroup' work?

simonnineteen90 replied:
yep

Kirsten1822 commented on test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js:
```diff
@@ -0,0 +1,46 @@
+
+const { orderGroupedEventsByScheme } = require('../../../../../app/data/events/scheme-id/order-grouped-events-by-scheme')
+const { SFI, SFI_PILOT, LUMP_SUMS, VET_VISITS, CS, BPS, FDMR } = require('../../../../../app/constants/schemes')
```
I'd put the constant at the top of the file and separate it with a space

Kirsten1822 commented on test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js:
```diff
@@ -0,0 +1,46 @@
+const { orderGroupedEventsByScheme } = require('../../../../../app/data/events/scheme-id/order-grouped-events-by-scheme')
+const { SFI, SFI_PILOT, LUMP_SUMS, VET_VISITS, CS, BPS, FDMR } = require('../../../../../app/constants/schemes')
+
+let groupedEvents
+let groupedEvent
```
I'd swap the order of these two around, since groupedEvents appears last below

Kirsten1822 commented on test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js:
```diff
@@ -0,0 +1,46 @@
+    groupedEvents = [bpsEvent, vetEvent, sfiEvent, sfiPilotEvent, lumpSumsEvent, csEvent, fdmrEvent]
+  })
+
+  test('should order grouped events into sequence by schemeId when all schemes present', () => {
```
What does 'by sequence' mean here?

Kirsten1822 replied:
Would schemeId be in ascending order then?

simonnineteen90 replied:
yep, have added that to test name

Kirsten1822 commented on test/unit/data/events/scheme/order-grouped-events-by-scheme.test.js:
```diff
@@ -0,0 +1,46 @@
+    expect(orderedGroupedEvents[6].schemeId).toEqual(FDMR)
+  })
+
+  test('should order grouped events into sequence by schemeId when only two schemes present', () => {
```
See my comment here: https://github.com/DEFRA/ffc-pay-data-hub/pull/2/files#r1212816728

simonnineteen90 replied:
updated to include ascending in description

Kirsten1822 commented on test/unit/data/events/scheme/sanitise-scheme-data.test.js:
```diff
@@ -0,0 +1,28 @@
+const totalSchemeValues = require('../../../../mocks/events/total-scheme-values')
+
+describe('get events by frn', () => {
+  beforeEach(() => {
+
+  })
```
Do you need this beforeEach if it's empty?

Kirsten1822 commented on test/unit/data/get-data.test.js:
```diff
@@ -26,11 +28,16 @@ describe('get data', () => {
   })
 
-  test('should get batch events for batch category', async () => {
+  test('should get schemeId events for scheme category', async () => {
     await getData(BATCH_CATEGORY, BATCH_VALUE)
     expect(mockGetEventsByBatch).toHaveBeenCalledWith(BATCH_VALUE)
   })
```
I'm not sure why this one was renamed here

simonnineteen90 commented on test/unit/data/events/scheme/get-total-scheme-values.test.js:
```diff
@@ -0,0 +1,76 @@
+    expect(result[0].value).toBe(300000)
+  })
+
+  test('value should be equal to sum of events values when there is only one event', () => {
```
there are multiple events for bps but i have removed two of the events with `.pop` for the first element (cs) of `groupedEvents` which is the part I am testing.

Kirsten1822 replied:
Then maybe it would be better to rename this test to '...only one event in first element of groupedEvents'?

simonnineteen90 commented:
have addressed comments and also changed name of function to `getEventsByScheme` as per Johns comment on the other repoo

Kirsten1822 commented on app/data/events/scheme-id/get-total-scheme-values.js:
```diff
@@ -0,0 +1,17 @@
+const getTotalSchemeValues = (groupedEvents) => {
+  return groupedEvents.map((eventsByScheme) => {
+    const value = eventsByScheme.events.reduce((total, current) => {
```
Thanks :)

Kirsten1822 commented:
> have addressed comments and also changed name of function to `getEventsByScheme` as per Johns comment on the other repoo

Will you need to change the test file names also?

simonnineteen90 commented:
> > have addressed comments and also changed name of function to `getEventsByScheme` as per Johns comment on the other repoo
> 
> Will you need to change the test file names also?

So some references to `schemeId` in the test names are correct as we are still using the `schemeId` for some tests but have been through and changed any that should just be `scheme`.

Kirsten1822 reviewed (approved):
Happy to approve :)
